[feat] refactor events to use ZMQ

andrejb opened the pull request:
Closes: #6359

kalikaneko commented:
requirements file should get protobuf & protobuf.socketrpc removed
```
protobuf>=2.4.1
protobuf.socketrpc
```

kalikaneko commented on src/leap/common/events/README.rst:
```diff
@@ -22,32 +25,29 @@ How to use it
-* Compile the new protocolbuffers file::
-
-  make
+To add a new event, just add it to ``signals.py``.
```
in the past, we've avoided the naming of "signals" so it cannot be confused with Qt signals.
maybe events.catalog or something like that?

andrejb replied:
Done.

andrejb replied:
Done. Moved `signal()` to `emit()` and `signals` to

kalikaneko commented on src/leap/common/events/signals.py:
```diff
@@ -0,0 +1,73 @@
+"""
+
+
+from enum import Enum
```
not important, but since there are some conflicts with enum/enum34 packaging in some distros, I was wondering if we could substitute it with simple ad-hoc implementation (given that we're not using so much of the enum api after all).

there is an example minimal implementation in twisted codebase.

ivanalejandro0 replied:
`enum34` is used by `cryptography`, so I would say we use `enum34` since we can't do much about it.
Is any place where we use `leap.common` and not `cryptography`? if there's a case maybe is a good idea to use an ad-hoc `enum`.

kalikaneko replied:
yes, I know it is used. but we're having troubles because (1) we have to backport it, (2) in some distros it might happen that some other package is requiring the installation of the old enum, which has a different api. all in all, I think we've spent more time dealing with packaging issues than the time it would have taken to code something simple. After all, we only need a simple class to wrap constants.

kalikaneko replied:
cryptography is only needed by pyOpenSSL >= 0.14 if I remember correctly, so it might be the case that in some distros it isn't there.

ivanalejandro0 replied:
> we're having troubles because (1) we have to backport it, 

we have to backport it, or not, depending on which `Cryptogrephy`/`pyOpenSSL` version we can/want to use... the problem will be there or not no matter if we discard this dependency for leap* code.

> (2) in some distros it might happen that some other package is requiring the installation of the old enum, which has a different api. 

in an old distro the problem will be which version of `Cryptogrephy`/`pyOpenSSL` we provide? because if we provide a source with bitmask, and a new `pyOpenSSL` which needs enum34 we will have problems anyway.

> all in all, I think we've spent more time dealing with packaging issues than the time it would have taken to code something simple. After all, we only need a simple class to wrap constants.

I agree, my point is not to use `enum34` or a custom one, but that we may want to stick with `enum34` since we will have problems with it anyway.

> cryptography is only needed by pyOpenSSL >= 0.14 if I remember correctly, so it might be the case that in some distros it isn't there.

Does this mean that discarding `enum`/`enum34` we can just ignore the problem and trust the distros to fix this issue for us?

kalikaneko replied:
> Does this mean that discarding enum/enum34 we can just ignore the problem and trust the distros to fix this issue for us? 

yes

> we have to backport it, or not, depending on which Cryptogrephy/pyOpenSSL version we can/want to use

right now, we depend on enum34, so we have had to backport it. in systems that also have enum we've got troubles.

kalikaneko replied:
> and a new pyOpenSSL which needs enum34 

when packaging debs, we're not providing pyopenssl, the distro does.
and bundles are not a problem since the lib scope is self-contained

ivanalejandro0 replied:
alright, I'm ok with an ad-hoc enum fix... at least until the distros we support drop the old `enum34`.

andrejb replied:
Done.

kalikaneko commented on src/leap/common/events/signals.py:
```diff
@@ -0,0 +1,73 @@
+    "IMAP_SERVICE_STARTED",
+    "IMAP_SERVICE_FAILED_TO_START",
+    "IMAP_CLIENT_LOGIN",
+    "IMAP_FETCHED_INCOMING",
```
this one should change to MAIL_FETCHED_INCOMING

andrejb replied:
Done.

kalikaneko commented on src/leap/common/events/signals.py:
```diff
@@ -0,0 +1,73 @@
+    "IMAP_MSG_PROCESSING",
+    "IMAP_MSG_DECRYPTED",
+    "IMAP_MSG_SAVED_LOCALLY",
+    "IMAP_MSG_DELETED_INCOMING",
```
these four above too

andrejb replied:
Done.

kalikaneko commented on src/leap/common/events/signals.py:
```diff
@@ -0,0 +1,73 @@
+    "IMAP_MSG_SAVED_LOCALLY",
+    "IMAP_MSG_DELETED_INCOMING",
+    "IMAP_UNHANDLED_ERROR",
+    "IMAP_UNREAD_MAIL",
```
and this one should be MAIL_UNREAD_MESSAGES(str, int) too

andrejb replied:
What do you mean by (str, int) ? Would these be arguments to the event? Currently, I added a simple "content" string, but I will add some kind of serialization to allow for *contents instead.

kalikaneko commented:
It would be a good idea to add the ascii diagrams somewhere in the repo as documentation (maybe a simple README, or even in the main events module docstrings).
Also, a little  explanation about the rationale of the change would be good in the body of the commit message.

kalikaneko commented on src/leap/common/events/client.py:
```diff
@@ -55,275 +54,269 @@ class CallbackAlreadyRegistered(Exception):
+        """
+        socket = self._context.socket(socktype)
+        stream = zmqstream.ZMQStream(socket, self._loop)
+        socket.connect("tcp://127.0.0.1:%d" % port)
```
Does this mean that any client can connect to the message channel?
We should try to authenticate the channel in the future.
Why not doing ipc under unix platforms for now?

ivanalejandro0 replied:
should we use CurveZMQ here and fallback to ipc with socket permissions as we did on the client?

kalikaneko commented on src/leap/common/events/client.py:
```diff
@@ -55,275 +54,269 @@ class CallbackAlreadyRegistered(Exception):
 
 
-def ensure_client_daemon():
+class EventsIOLoop(ioloop.ZMQIOLoop):
```
does this mean that all clients will use the zmqioloop?
isn't this only needed for the qt frontend?

kalikaneko replied:
ah, I see now that txzmq client is not implemented yet.

kalikaneko commented on src/leap/common/events/client.py:
```diff
@@ -55,275 +54,269 @@ class CallbackAlreadyRegistered(Exception):
+                self._loop.stop(wait=True)
+
+
+_client = EventsClient.instance()
```
I don't quite get the reason for the existence of this singleton

ivanalejandro0 replied:
hmm, using module-singleton behavior to create the client once?

andrejb replied:
The idea of the singleton is to save ZMQ connections and memory by having all code using events to use the same client. We want client object and connection to be transparent to the user of events (i.e. the api user doesn't have to explicitelly instantiate the connection), so the other possibility would be to use one client object and two ZMQ connections per request, which would add some overhead for each event sent.

There is a problem, though, with current (eventloop) implementation, which is that the user would have to explicitelly call `client.shutdown()` to terminate the client thread, otherwise the program will not finish. I can't see other options for the events api user than the ones below, maybe you can comment and add others:

* Use one client object and/or zmq connection for each message sent.
* Manage the connection shutdown explicitelly.
* Use txzmq twisted-reactor-powered events client (yet to be implemented) and schedule the connection termination to be made on reactor shutdown.

ivanalejandro0 replied:
I'm ok with the singleton variable in there.
You don't really need it though, you can use `EventsClient.instance()` in each place when you use `_client` and will behave the same way since the `instance` method 'singletonify' that for you hehe.

kalikaneko replied:
> There is a problem, though, with current (eventloop) implementation, which is that the user would have to explicitelly call client.shutdown()

I would give a try to atexit.register(...) https://docs.python.org/2/library/atexit.html

kalikaneko replied:
> hmm, using module-singleton behavior to create the client once?

to do this, you don't need to instantiate the object on each import. it might suffice to keep a module-level container.

kalikaneko commented on src/leap/common/events/server.py:
```diff
@@ -24,213 +24,94 @@
     """
-    Service for receiving events in clients.
+    An events server that listens for signals in one port and publishes those
+    signals in another port.
```
refer to my note above about the use of the term "signals" instead of "events"

andrejb replied:
Done.

kalikaneko commented on src/leap/common/events/server.py:
```diff
@@ -24,213 +24,94 @@
-        done.run(response)
-
-    def ping(self, controller, request, done):
+        self._pub = self._zmq_bind(txzmq.ZmqPubConnection, port)
```
shouldn't it capture the "already binded port" kind of error?

andrejb replied:
Previous server used to do this. I removed it because I thought that making sure there's only one server running should be responsibility of the higher level code using the events server api. This would be reinforced by the fact that we're now using the twisted reactor to manage the events server.

Do you think we should re-implement this here, in the server level?

kalikaneko replied:
hmm maybe it's good to leave the exception catching to the caller. I was more worried about the possibility that the port is already in use by another application.

kalikaneko commented on src/leap/common/events/server.py:
```diff
@@ -24,213 +24,94 @@
-logger = logging.getLogger(__name__)
 
+import logging
+import txzmq
```
txzmq needs to be added to requirements.

andrejb replied:
Done.

ivanalejandro0 commented on src/leap/common/events/client.py:
```diff
@@ -55,275 +54,269 @@ class CallbackAlreadyRegistered(Exception):
-    returned for a synch request but nothing will be returned for an asynch
-    request.
+    @classmethod
+    def instance(self):
```
I think that the preferred name for this parameter when you use `@classmethod` is `cls` (see: https://docs.python.org/2/library/functions.html#classmethod) instead of `self`.

andrejb replied:
You're correct. :-)

ivanalejandro0 commented on src/leap/common/events/client.py:
```diff
@@ -55,275 +54,269 @@ class CallbackAlreadyRegistered(Exception):
+        Return a singleton EventsClient instance.
+        """
+        with EventsClient._instance_lock:
+            if not EventsClient._instance:
```
I think this should be:
```python
if EventsClient._instance is not None:
```

andrejb replied:
Done.

ivanalejandro0 commented:
According to https://github.com/leapcode/bitmask_client/blob/develop/docs/leap-commit-template
Instead of `[feature] Refactor leap.common.events to use ZMQ.` the commit message should be `[feat] refactor leap.common.events to use ZMQ`
The Capital case and the trailing dot are not documented how we wanted yet, I prefer the latter (even though some time ago I did any of the two depending on my preference that day hehe) and @kalikaneko rolls a dice each time. Do you prefer capitalized and with a trailing dot?

andrejb commented:
Fixed requirements, added the ascii art diagram to readme, added comment about the change to commit message.

About commit message style, for me uppercase with trailing dot reminds a phrase, while lowercase and no trailing dot reminds a title. Strangelly, even though i find the former more "correct", i also find the latter more cute and easier to read.

Still have to do:

* txzmq client
* serialization of multiple arguments to `emit()`.
* decide if will use, in pure zmq client, one connection per event emition, or explicitelly handle connection shutdown, or anything else.
* decide if will handle port already taken for server binding.

kalikaneko commented on src/leap/common/events/README.rst:
```diff
@@ -22,32 +86,25 @@ How to use it
 
 >>> from leap.common.events import server
->>> server.ensure_server(port=8090)
+>>> server.ensure_server(sig_port=9000, pub_port=9001)
```
evt_port?

kalikaneko commented on src/leap/common/events/catalog.py:
```diff
@@ -0,0 +1,88 @@
+]
+
+
+class Event:
```
Event(object)

kalikaneko commented on src/leap/common/events/client.py:
```diff
@@ -55,275 +55,276 @@ class CallbackAlreadyRegistered(Exception):
+    Shutdown the events client thread.
+    """
+    EventsClient.instance().shutdown()
+
```
If automatically shutting down is a concern, I'd give a try at atexit.register() https://docs.python.org/2/library/atexit.html

kalikaneko commented on src/leap/common/events/catalog.py:
```diff
@@ -0,0 +1,88 @@
+    "SOLEDAD_UPLOADING_KEYS",
+    "UPDATER_DONE_UPDATING",
+    "UPDATER_NEW_UPDATES",
+]
```
just a disgression, but... in the future we could have a decentralized register of events.
each module would announce the events that it can signal, so that they can be discovered by other modules.

ivanalejandro0 commented:
We will need to bump this to 0.4 for the new api, right?

kalikaneko commented:
bumping to 0.4 sounds right

ivanalejandro0 commented:
We have these differences between the old and new events:

```diff
diff --git a/old.txt b/new.txt
index e5f8648..499e20b 100644
--- a/old.txt
+++ b/new.txt
@@ -1,21 +1,21 @@
 CLIENT_SESSION_ID
 CLIENT_UID
 IMAP_CLIENT_LOGIN
-IMAP_FETCHED_INCOMING
-IMAP_MSG_DECRYPTED
-IMAP_MSG_DELETED_INCOMING
-IMAP_MSG_PROCESSING
-IMAP_MSG_SAVED_LOCALLY
 IMAP_SERVICE_FAILED_TO_START
 IMAP_SERVICE_STARTED
 IMAP_UNHANDLED_ERROR
-IMAP_UNREAD_MAIL
 KEYMANAGER_DONE_UPLOADING_KEYS
 KEYMANAGER_FINISHED_KEY_GENERATION
 KEYMANAGER_KEY_FOUND
 KEYMANAGER_KEY_NOT_FOUND
 KEYMANAGER_LOOKING_FOR_KEY
 KEYMANAGER_STARTED_KEY_GENERATION
+MAIL_FETCHED_INCOMING
+MAIL_MSG_DECRYPTED
+MAIL_MSG_DELETED_INCOMING
+MAIL_MSG_PROCESSING
+MAIL_MSG_SAVED_LOCALLY
+MAIL_UNREAD_MESSAGES
 RAISE_WINDOW
 SMTP_CONNECTION_LOST
 SMTP_END_ENCRYPT_AND_SIGN
```

It seems like `leap.mail` doesn't implement them yet and bitmask fails because of a non existent `IMAP_UNREAD_MAIL`. Should I change to those new names in https://github.com/leapcode/leap_mail/pull/178 ?

kalikaneko commented:
@ivanalejandro0 yes, IMAP_MSG_* should be moved to MAIL_MSG_*

ivanalejandro0 commented on src/leap/common/events/catalog.py:
```diff
@@ -0,0 +1,88 @@
+    "KEYMANAGER_KEY_NOT_FOUND",
+    "KEYMANAGER_LOOKING_FOR_KEY",
+    "KEYMANAGER_STARTED_KEY_GENERATION",
+    "MAIL_FETCHED_INCOMING",
```
@kalikaneko @andrejb -- does this name change (`IMAP_FETCHED_INCOMING` -> `MAIL_FETCHED_INCOMING`) is correct?

kalikaneko replied:
yes, it is. the fetch functionality has been moved to its own place for mail 0.4, it doesn't belong to imap anymore.

ivanalejandro0 replied:
cool, added to mail changes.

ivanalejandro0 commented on src/leap/common/events/client.py:
```diff
@@ -21,309 +23,497 @@
-    """
-    Ensure the client daemon is running and listening for incoming
-    messages.
+class EventsClient(object):
```
Please add some docstring to this.

ivanalejandro0 commented on src/leap/common/events/client.py:
```diff
@@ -21,309 +23,497 @@
+        print self._stopped
+
+
+class EventsClientThread(threading.Thread, EventsClient):
```
Please add some docstring to this.

meskio commented on src/leap/common/tests/test_events.py:
```diff
@@ -15,414 +15,156 @@
+        self._server = server.ensure_server(
+            emit_addr="tcp://127.0.0.1:0",
+            reg_addr="tcp://127.0.0.1:0")
+        self._client.configure_client(
```
I don't see where this `self._client` is initializated, it looks like it should be `self._client = client.configure_client(`. But weirdly I get an unrelated error by running the test:
```
Failure: twisted.trial.util.DirtyReactorAggregateError: Reactor was unclean.
Selectables:
ZmqPubConnection(ZmqFactory(), [])
ZmqPullConnection(ZmqFactory(), [])
ZmqPullConnection(ZmqFactory(), [])
ZmqPubConnection(ZmqFactory(), [])

leap.common.tests.test_events.EventsClientTestCase.test_client_receives_signal
```

meskio replied:
Never mind my comment. I see now that `_client` is set by the inheriting class. The error I was seeing I'm not sure if is my fault, I'm doing some experiments with the tests.

meskio commented on src/leap/common/tests/test_events.py:
```diff
@@ -15,414 +15,156 @@
-    def tearDownClass(cls):
-        # give some time for requests to be processed.
-        time.sleep(1)
+class EventsGenericClientTestCase(object):
```
This tests generate `~/.config/leap/events` folder. Our tests usually try to don't write in your HOME folder, see `leap.common.testing.basetest.BaseLeapTest`.

I don't think this is a blocker but might be a good idea to add a comment with a `XXX` to be fixed in the future.
